### PR TITLE
Add Pedersen gadget

### PR DESCRIPTION
**Pull request by jules**

### CODE REVIEW

File: circuits/core/src/pedersen/hash.rs
Comment (howardwu) [inline circuits/core/src/pedersen/hash.rs]: This should be a ternary, since the hash is hardcoded in the circuit otherwise.

**Review by howardwu — CHANGES_REQUESTED** [review]

File: circuits/core/src/pedersen/hash.rs
Comment (howardwu) [inline circuits/core/src/pedersen/hash.rs]: Let's move this to a `check_setup` and `test_setup_constant` function in `mod.rs`

File: circuits/core/src/pedersen/hash.rs
Comment (howardwu) [inline circuits/core/src/pedersen/hash.rs]: This should be moved out of `scope` otherwise the constraint counts are incorrect (as they will factor these into the count)

File: circuits/core/src/pedersen/hash.rs
Comment (howardwu) [inline circuits/core/src/pedersen/hash.rs]: I think `.sum()` might work here

Comment (jules) [reply circuits/core/src/pedersen/hash.rs]: It wouldn't let me, apparently `Iterator::Sum` wasn't implemented for `Group`

Comment (howardwu) [reply circuits/core/src/pedersen/hash.rs]: That's helpful, I will add that in a future PR.

File: circuits/core/src/pedersen/hash.rs
Comment (howardwu) [inline circuits/core/src/pedersen/hash.rs]: Out of curiosity, does `&base.clone()` here need a `.clone()`? I thought pass by reference would be sufficient. (I assume its needed since Clippy CI didn't seem to complain about it)

**Review by howardwu — APPROVED** [review]
LGTM, let's wait to merge this after the base branch is merged into `testnet3`

File: circuits/core/src/pedersen/hash.rs
Comment (jules) [reply circuits/core/src/pedersen/hash.rs]: You're absolutely right, I don't need it, will remove it

File: circuits/core/src/pedersen/hash.rs
Comment (ljedrz) [inline circuits/core/src/pedersen/hash.rs]: it the `collect` necessary here? `flat_map` might work without it

Comment (jules) [reply circuits/core/src/pedersen/hash.rs]: I couldn't get it to work with `flat_map` as it was complaining `Group` wasn't an iterator

**Pull request merged**
